### PR TITLE
add csharp support

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

Add c# api

## TODO
- [x] static/shared build runtime test
- [x] nuget package
- [x] code convention check (StyleCop.Analyzers)
- [x] document.

### CODE REVIEW

File: csrc/apis/csharp/src/MMDeploySharp/APIs/Classifier.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace MMDeploySharp
```
Comment (lvhan028) [inline csrc/apis/csharp/src/MMDeploySharp/APIs/Classifier.cs:4]: I think `namespace mmdeploy {namespace csharp {}}` is better

File: csrc/apis/csharp/src/MMDeploySharp/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+[assembly: AssemblyCopyright("Copyright ©  2022")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+// Setting ComVisible to false makes the types in this assembly not visible
+// to COM components.  If you need to access a type in this assembly from
+// COM, set the ComVisible attribute to true on that type.
+[assembly: ComVisible(false)]
+
+// 如果此项目向 COM 公开，则下列 GUID 用于类型库的 ID
```
Comment (lvhan028) [inline csrc/apis/csharp/src/MMDeploySharp/Properties/AssemblyInfo.cs:21]: Better to use English comments

File: csrc/apis/csharp/src/MMDeploySharp/APIs/Classifier.cs (hunk 1 above)
Comment (irexyc) [reply csrc/apis/csharp/src/MMDeploySharp/APIs/Classifier.cs:3]: Found It is common that package name is same with root namespace.

File: csrc/apis/csharp/src/MMDeploy/APIs/common.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
+﻿#pragma warning disable 1591
+namespace MMDeploy
+{
+    /// <summary>
+    /// Pixel format.
+    /// </summary>
+    public enum MmPixelFormat
```
Comment (lvhan028) [inline csrc/apis/csharp/src/MMDeploy/APIs/common.cs:7]: As discussed in the meeting, remove the prefix 'Mm'

File: csrc/apis/csharp/src/MMDeploy/APIs/Classifier.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
+﻿using System.Collections.Generic;
+
+namespace MMDeploy
+{
+    /// <summary>
+    /// Single classification result of a picture.
+    /// A picture may contains multiple reuslts.
+    /// </summary>
+    public struct MmClass
```
Comment (lvhan028) [inline csrc/apis/csharp/src/MMDeploy/APIs/Classifier.cs:9]:
```csharp
public struct Label {
    public int Id;
    public float score;
}
```

File: csrc/apis/csharp/src/MMDeploy/APIs/DisposableObject.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    public class DisposableObject : IDisposable
+    {
+#pragma warning disable SA1401 // Fields should be private
+        /// <summary>
+        /// Handle pointer.
+        /// </summary>
+        protected IntPtr _handle;
+#pragma warning restore SA1401 // Fields should be private
+
+        private bool disposed = false;
```
Comment (lvhan028) [inline csrc/apis/csharp/src/MMDeploy/APIs/DisposableObject.cs:18]: inconsistent naming rule: `_handle` vs `disposed`

**Review by lvhan028 — COMMENTED** [review]
Action items:
1. Follow the C# name convention. @lzhangzz , please provide the spec and highlight the crucial rules.
2. Remove the prefix 'Mm' or 'CMm' of the data structure, such as `MmMat`, `MmPointf`, etc.
3. Eliminate the absolute path in demo codes.
4. Rename the widget name in the image classification demo. ``xxx{digit}` is not allowed.

**Comment by lzhangzz** [comment]
https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions

File: csrc/apis/csharp/MMDeploy/APIs/Classifier.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        private unsafe void FormatResult(int matCount, int* resultCount, Label* results, ref List<ClassifierOutput> output, out int total)
+        {
+            total = 0;
+            for (int i = 0; i < matCount; i++)
+            {
+                ClassifierOutput outi = default;
+                for (int j = 0; j < resultCount[i]; j++)
+                {
+                    outi.Add(results[j].Id, results[j].Score);
+                    results++;
```
Comment (lvhan028) [inline csrc/apis/csharp/MMDeploy/APIs/Classifier.cs:116]: Doesn't `results++` result in an out-of-boundary issue, does it? In L115, it is visited by index.

**Review by lvhan028 — DISMISSED** [review]

File: CMakeLists.txt [hunk 7]
```diff
@@ -28,6 +28,7 @@ set(MMDEPLOY_TARGET_DEVICES
         "cpu" CACHE STRING "MMDeploy's target devices")
 set(MMDEPLOY_TARGET_BACKENDS "" CACHE STRING "MMDeploy's target inference engines")
 set(MMDEPLOY_CODEBASES "all"  CACHE STRING "select OpenMMLab's codebases")
+set(MMDEPLOY_BUILD_CSHARP_EXTERN "build MMDeployExtern.dll for c# api" OFF)
```
Comment (lzhangzz) [inline CMakeLists.txt:31]: To keep the naming consistent, use `MMDEPLOY_BUILD_SDK_CSHARP_API`, and it's "C# API"

**Review by lzhangzz — APPROVED** [review]

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**